### PR TITLE
WIP - Add quota and apispec to mixerclient filter configuration

### PR DESCRIPTION
**Pull request by ayj**
**What this PR does / why we need it**:

Add scaffolding for including HTTPAPISpec and QuotaSpec in mixerclient filter configuration. 

TODO 
* `HTTPAPISpecByDestination` and `QuotaSpecByDestination` configuration store implementation.
* `SortHTTPAPISpec` and `SortQuotaSpec` implementation to avoid unnecessary config churn.
* Validation functions

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by ayj** [comment]
Posting this for early review before I finished fleshing out the implementation.

/cc @rshriram @kyessenov

**Comment by ayj** [comment]
This PR includes the full diff from https://github.com/istio/pilot/pull/1644. See commit https://github.com/istio/pilot/pull/1661/commits/e8154de2b930d4b6b93f63e5dc1b9eff98fb7de4 for the  specific changes to add quota and apispec to mixerclient filter config.

**Comment by rshriram** [comment]
Closing.. code moved to istio/istio

File: proxy/envoy/discovery.go [hunk 1]
```diff
@@ -226,6 +226,11 @@ func NewDiscoveryService(ctl model.Controller, configCache model.ConfigStoreCach
 		configCache.RegisterEventHandler(model.IngressRule.Type, configHandler)
 		configCache.RegisterEventHandler(model.EgressRule.Type, configHandler)
 		configCache.RegisterEventHandler(model.DestinationPolicy.Type, configHandler)
+
+		configCache.RegisterEventHandler(model.HTTPAPISpec.Type, configHandler)
+		configCache.RegisterEventHandler(model.HTTPAPISpecBinding.Type, configHandler)
+		configCache.RegisterEventHandler(model.QuotaSpec.Type, configHandler)
+		configCache.RegisterEventHandler(model.QuotaSpecBinding.Type, configHandler)
```
Comment (rshriram) [inline proxy/envoy/discovery.go]: can you add a note here saying that this invalidation is too coarse grained? API spec changes result in mixer filter changes. But it does not mean we have to blow up everything in the cache.

File: proxy/envoy/mixer.go [hunk 2]
```diff
@@ -168,24 +171,50 @@ func mixerHTTPRouteConfig(role proxy.Node, services []string) *FilterMixerConfig
...
+		var services []string
+		if instances != nil {
+			serviceSet := make(map[string]bool, len(instances))
+			for _, instance := range instances {
+				serviceSet[instance.Service.Hostname] = true
+			}
+			for service := range serviceSet {
+				services = append(services, service)
+			}
+			sort.Strings(services)
```
Comment (rshriram) [inline proxy/envoy/mixer.go]: stablesort?

**Pull request closed without merging**